### PR TITLE
Fix datetime decoding when time units are 'days since 0000-01-01 00:00:00'

### LINKED ISSUE #521: time decoding error with "days since"
I am trying to use xray with some CESM [POP model netCDF output](http://www.cesm.ucar.edu/models/ccsm3.0/pop/doc/POPusers_chap4.html), which supposedly follows CF-1.0 conventions. It is failing because the models time units are "'days since 0000-01-01 00:00:00". When calling open_dataset, I get the following error:
```
ValueError: unable to decode time units u'days since 0000-01-01 00:00:00' with the default calendar. Try opening your dataset with decode_times=False. Full traceback:
Traceback (most recent call last):
  File "/home/rpa/xray/xray/conventions.py", line 372, in __init__
    # Otherwise, tracebacks end up swallowed by Dataset.__repr__ when users
  File "/home/rpa/xray/xray/conventions.py", line 145, in decode_cf_datetime
    dates = _decode_datetime_with_netcdf4(flat_num_dates, units, calendar)
  File "/home/rpa/xray/xray/conventions.py", line 97, in _decode_datetime_with_netcdf4
    dates = np.asarray(nc4.num2date(num_dates, units, calendar))
  File "netCDF4/_netCDF4.pyx", line 4522, in netCDF4._netCDF4.num2date (netCDF4/_netCDF4.c:50388)
  File "netCDF4/_netCDF4.pyx", line 4337, in netCDF4._netCDF4._dateparse (netCDF4/_netCDF4.c:48234)
ValueError: year is out of range
```

Full metadata for the time variable:
```
	double time(time) ;
		time:long_name = "time" ;
		time:units = "days since 0000-01-01 00:00:00" ;
		time:bounds = "time_bound" ;
		time:calendar = "noleap" ;
```

I guess this is a problem with the underlying netCDF4 num2date package?

### PR DESCRIPTION
**Pull request by rabernat**
This fixes #521 using the workaround described in Unidata/netcdf4-python#442.

### CODE REVIEW

**Comment by jhamman** [comment]
I have a few general comments,  

1.  Can you point me to where in the [CF Conventions](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.6/build/cf-conventions.html#time-coordinate) or [UDUNITS](http://www.unidata.ucar.edu/software/udunits/#documentation) the valid time coordinate units defined?
2.  We should think about whether or not this fix belongs in xray or netCDF4.  I am of the opinion that if the CF Conventions do in fact support the units in question, we should apply this fix in the netCDF package. If they don't, I don't think we wan to support it here either.
3.  @shoyer is out for a bit but he will almost certainly want to weigh in on this.

If we end up going this route, you'll want to add some unit tests in [test_conventions.py](https://github.com/xray/xray/blob/master/xray/test/test_conventions.py#L234-L254).

File: xray/conventions.py [hunk 1]
```diff
@@ -94,7 +94,20 @@ def _unpack_netcdf_time_units(units):
...
-    dates = np.asarray(nc4.num2date(num_dates, units, calendar))
+    try:
+        dates = np.asarray(nc4.num2date(num_dates, units, calendar))
+    except ValueError:
+        # num2date needs calendar year start >= 0001 C.E. (bug submitted
+        # to unidata about this)
+        since_yr_idx = units.index('since ') + 6
+        year = int(units[since_yr_idx:since_yr_idx+4])
+        assert year==0, 'this fix only works for days since year 0'
+        year_diff = year - 0001
```
Comment (jhamman) [inline xray/conventions.py]: Python 3 is raising a `SyntaxError` here.  This should just be `year_diff = year - 1`.

File: xray/conventions.py [hunk 2]
```diff
@@ -94,7 +94,20 @@ def _unpack_netcdf_time_units(units):
...
 
-    dates = np.asarray(nc4.num2date(num_dates, units, calendar))
+    try:
+        dates = np.asarray(nc4.num2date(num_dates, units, calendar))
+    except ValueError:
+        # num2date needs calendar year start >= 0001 C.E. (bug submitted
+        # to unidata about this)
+        since_yr_idx = units.index('since ') + 6
+        year = int(units[since_yr_idx:since_yr_idx+4])
+        assert year==0, 'this fix only works for days since year 0'
```
Comment (jhamman) [inline xray/conventions.py]: I don't really like the syntax of this assertion. Better to use an if statement and raise an exception:

```python
if year == 0:
    raise ValueError('this fix only works for days since year 0')
```

File: xray/conventions.py [hunk 3]
```diff
@@ -94,7 +94,20 @@ def _unpack_netcdf_time_units(units):
...
+        year_diff = year - 0001
+        new_units = units[:since_yr_idx] + '0001-01-01 00:00:00'
```
Comment (jhamman) [inline xray/conventions.py]: This will not work if the beginning month/day/hour/minute/second is not `01-01 00:00:00`

File: xray/conventions.py [hunk 4]
```diff
@@ -94,7 +94,20 @@ def _unpack_netcdf_time_units(units):
...
     import netCDF4 as nc4
 
-    dates = np.asarray(nc4.num2date(num_dates, units, calendar))
+    try:
+        dates = np.asarray(nc4.num2date(num_dates, units, calendar))
+    except ValueError:
+        # num2date needs calendar year start >= 0001 C.E. (bug submitted
+        # to unidata about this)
+        since_yr_idx = units.index('since ') + 6
+        year = int(units[since_yr_idx:since_yr_idx+4])
```
Comment (jhamman) [inline xray/conventions.py]: maybe just use 

```python
_, ref_date = _unpack_netcdf_time_units(units)
year = int(ref_date.split('-')[0])
```

**Comment by ocefpaf** [comment]
> Can you point me to where in the CF Conventions or UDUNITS the valid time coordinate units defined?

This edge case is defined here: http://cfconventions.org/Data/cf-conventions/cf-conventions-1.6/build/cf-conventions.html#climatological-statistics

NB: People that follow the COARDS conventions will also need this.  Any Udunits wrapper can deal with that for you:

```python
import cf_units
u = cf_units.Unit('days since 0000-01-01 00:00:00', calendar=cf_units.CALENDAR_NO_LEAP)
ut = u.utime()
ut.num2date(0)  # Returns a fake datetime object (http://scitools.org.uk/iris/docs/latest/iris/iris/unit.html#iris.unit.Unit.num2date)
0-01-01 00:00:00
ut.date2num(datetime(1, 1, 1))  # Note that datetime cannot take year = 0
365.0
```

> We should think about whether or not this fix belongs in xray or netCDF4. I am of the opinion that if the CF Conventions do in fact support the units in question, we should apply this fix in the netCDF package. If they don't, I don't think we wan to support it here either.

I strongly disagree with that.  netCDF is not **one** CDM and should not follow the CF-conventions!  I know that iris does follow CF very closely (annoyingly in fact) and xray "kind of" follow (which is great BTW).  However, if such conventions were adopted in the netCDF package, how will we load non-CF files?

**Comment by ocefpaf** [comment]
Note that I am not question the validity of those dates neither its use.  But they are out there and people need to deal with that.  If xray decides to support this it has to be implemented very carefully and with some warnings...

**Comment by rabernat** [comment]
@ocefpaf NCAR is one of the lead institutions in terms of the CF conventions. Yet the CESM POP model, also developed at NCAR, has this "year 0" issue! To me this suggests that any practical, real-world application will need to deal with special cases like this one.

**Comment by ocefpaf** [comment]
> @ocefpaf NCAR is one of the lead institutions in terms of the CF conventions. Yet the CESM POP model, also developed at NCAR, has this "year 0" issue! To me this suggests that any practical, real-world application will need to deal with special cases like this one.

I know :wink: that is why I use `cf_units`!

PS: do you have an OPenDAP endpoint for you data?  I would like to make a few tests here.

**Comment by rabernat** [comment]
@ocefpaf Unfortunately no OPenDAP endpoint, but I put a sample file on our ftp server
ftp://ftp.ldeo.columbia.edu/pub/rpa/pop_sample/b40.1850.track1.2deg.wcm.007.pop.h.0100-01.nc

**Comment by ocefpaf** [comment]
@rabernat here is an example on how to read those dates using `cf_units` (big download for a small example :wink:):

http://nbviewer.ipython.org/gist/ocefpaf/d14bd8ad24f4e1a47b19

My point with this is example is:  `netCDF4-python` should not do that for you!  And, of `xray` wants to go down that road, it must be done with some warnings and good documentation!

By accepting these rules (by rules I mean how UDUNITS interpret the dates and that this is part of the CF-convention CDM) you are leaving the world of "read my data" and entering a world of "interpret my data."

`cf_units` interprets your data as CF-compliant and does what you expect from that CDM.

File: xray/conventions.py [hunk 5]
```diff
@@ -94,7 +94,26 @@ def _unpack_netcdf_time_units(units):
...
+        if year != 0:
+            raise ValueError('This fix only works for days since year 0.')
+        year_diff = year - 1
+        new_units = re.sub('0000', '0001', units)
+        time = nc4.num2date(num_dates, new_units, calendar)
+        dates = np.asarray([datetime(d.year + year_diff, d.month, d.day,
+                         d.hour, d.minute, d.second)
+                for d in time])
+        warnings.warn('Detected time units with reference date of year 0. '
+                      'This is not valid according to CF convetions. '
```
Comment (ocefpaf) [inline xray/conventions.py]: It is valid!  It is not recommended.

**Comment by rabernat** [comment]
Here is another example of a dataset with a similar time encoding issue. This is a valid OpenDAP URL:
http://data.nodc.noaa.gov/thredds/dodsC/woa/WOA13/DATAv2/temperature/netcdf/decav/1.00/woa13_decav_t16_01v2.nc

The time units are 'months since 0000-01-01 00:00:00'. This is the very popular [World Ocean Atlas](https://www.nodc.noaa.gov/OC5/woa13/). Since it is a monthly climatology, the dates are referenced to year 0.

**Comment by shoyer** [comment]
I would also strongly prefer to get this fix in netCDF4 upstream rather than in xray, if possible.

**Comment by ocefpaf** [comment]
> I would also strongly prefer to get this fix in netCDF4 upstream rather than in xray, if possible.

I disagree.  Maybe not even xray should "fix" this.  There are two issues here:

1) Interpreting year 0 as 1;
2) Non-standard calendars.

Maybe (2) should be in `np.datetime64` (and pandas), but even that is hard to became a reality.  However, (1) is an UDUNITS ~~aberration~~ interpretation that made into some Conventions (CF and [COARDS](http://ferret.wrc.noaa.gov/noaa_coop/coop_cdf_profile.html)).  Does xray wants to became a CF CDM?  I sure hope not!  I have iris for that but for **everything** else I use xray.

And if people really want to get the dates into xray they can just read the raw time data and parse it using  one of the several UDUNITS wrapper or convention checker/compliance out there. Here is another one that should do the right thing here: https://pypi.python.org/pypi/coards/1.0.5

**Comment by ocefpaf** [comment]
Just so you see how messy this can be.  The year `0000`, because it does not exist, is used to store climatology in COARDS and CF.  The `coards` package issued a warning, but did the wrong thing:

```python
from coards import parse
units = 'days since 0000-01-01 00:00:00'
parse(0, units)
coards/__init__.py:60: UserWarning: Shifted data 366 days to the future, since year zero does not exist.
  UserWarning)
datetime.datetime(1, 1, 1, 0, 0)
```

If someone accidentally saves that date back with this object it will no longer be year 0 and other CDMs might choke by not recognizing it as climatology.  `cf_units` does the right thing:

```python
import cf_units
u = cf_units.Unit('days since 0000-01-01 00:00:00', calendar=cf_units.CALENDAR_NO_LEAP)
ut = u.utime()
ut.num2date(0)
0-01-01 00:00:00
```

But that object is pretty much useless for pandas and xray because of the non-standard calendar.  The object **is** useful to annotate plots or to save the data metadata back the same way it was before.

**Comment by shoyer** [comment]
@ocefpaf To be clear, by "strongly prefer to get this fix upstream" I mostly meant that I am reluctant to include this in xray.

I *would* like it to be straightforward for others to extend our reading capabilities for netcdfs by adding custom logic like this for their own equivalent of `xray.open_dataset` that builds on what we have in xray.

**Comment by ocefpaf** [comment]
> @ocefpaf To be clear, by "strongly prefer to get this fix upstream" I mostly meant that I am reluctant to include this in xray.

Great.  We don't need another interpretation of the CF-standards out there!

See the [WOA13](http://data.nodc.noaa.gov/thredds/dodsC/woa/WOA13/DATAv2/temperature/netcdf/decav/1.00/woa13_decav_t16_01v2.nc.html) dataset above as an example of the problems this brings. The data has no calendar and use units `months since`.  CF discourage the use of [months](https://github.com/Unidata/netcdf4-python/issues/434) for obvious reasons and no calendar translates to:

>  gregorian or standard Mixed Gregorian/Julian calendar as defined by Udunits.

Which one!  Year zero is invalid in one and (kind of) accepted in the other!!

Long story short: the dataset is valid CF but it is unclear how to interpret the dates.  No one cares because people using that dataset know that `t16` means Autumn (Oh no wait!! I am in the South Hemisphere, so that means Spring :stuck_out_tongue_winking_eye:)

PS: Note that there is a `none` calendar too in the CF conventions that is more adequate for datasets like that.

**Comment by rabernat** [comment]
Ok, thanks for the thoughtful discussion. I understand why you both feel this shouldn't be implemented in xray. My one objection to the discussion is that I don't think that climatological time is such a "niche" issue--processing climate model output (much of which has no specific calendar date associated with it but still has seasonal cycles, etc.) seems like one of the most useful applications for xray.

I come at this from a very practical point of view. I _need_ to use certain datasets (e.g. WOA13, POP model output) for my research and teaching. I _want_ to teach xray in my fall physical oceanography class---it is perfect for teaching because it "just works" and allows students to quickly load data and do basic analysis easily. This PR request would allow me to do that. I understand the counter-arguments, but I don't know exactly how I should proceed if this can't be part of xray. The options seem to be:

1) Don't use the datasets
2) Don't use xray
3) Use xray without time support

Is there something else I'm missing?

**Comment by ocefpaf** [comment]
> I want to teach xray in my fall physical oceanography class

@rabernat I hear you!  I suffer from the same problem.  But I think we should teach students how to defend themselves from bad practices and tools limitations.  (We have both in there!  Talking about non-standard calendar and accepted but non-recommended standards.)

> 3) Use xray without time support

Not ideal but that is probably the way to go.  My guess is that you can load the data in xray, but you cannot get those dates into the pandas indexing machinery, right? That is not too bad since there are not many "dates operations" that you can do anyway.  Most of the time we only need to translate the time information into a label for the figures. If that is the case you are OK with the tools we have now.

**Comment by rabernat** [comment]
Regarding the non-standard calendar support, is it worth opening issues in numpy / pandas?

**Comment by ocefpaf** [comment]
> Regarding the non-standard calendar support, is it worth opening issues in numpy / pandas?

Worth? Yes.  Any hope to actually get it in there?  No...

**Comment by jhamman** [comment]
>Worth? Yes. Any hope to actually get it in there? No...

I think I disagree.  There is almost no chance anyone outside of the climate community is going to spend time on this but, if calendar support was added in a responsible way to numpy and pandas, I don't see why they wouldn't be interested.  So it will need to come from the climate users community, which IMHO, is under represented in the dev community.

If you don't want open the issue, I will.

**Comment by ocefpaf** [comment]
@jhamman sorry for making this thread longer than it should.  But I don't think you disagree! You are just more optimist than I am  :wink:

But you are right.  We need a champion from the climate community.  And if either of you open the issue I will be the [second man](https://www.youtube.com/watch?v=fW8amMCVAJQ) on the hill.

**Comment by shoyer** [comment]
Calendar support is numpy is conceivable, but it will pretty much require fixing numpy dtypes first so that they can be parametrized and extended by third parties in Python (this is on the roadmap). Right now the datetime64 type itself is pretty buggy, in large part because it's written in C code that nobody is maintaining.

.

For pandas, I think the bigger issue is that pandas only does datetime64 with ns resolution. Simply adding us support would go a long ways toward solving this. See here for some discussion on the pandas side: https://github.com/pydata/pandas/issues/7307

**Comment by rabernat** [comment]
I am the least knowledgeable person here regarding numpy and pandas development. The issues should probably be opened by someone else.

**Comment by jhamman** [comment]
I submitted an feature request over at numpy.  I'm going to close this now.

@rabernat - thanks for the PR and keep them coming.
